File: exotic-options/asian_options/monte-carlo-method/arithmetic_asian.py
"""
A script to simulate arithmetic Asian options using Monte-Carlo simulations

Author: Harsh Parikh
Date: 03rd March 2025
"""

import numpy as np


def arithmetic_asian_option(
    spot, strike, sigma, T, r, M=5000, N=5000, option_type="call"
):
    """
    Prices an arithmetic Asian option using Monte Carlo simulation under the
    Geometric Brownian Motion (GBM) model.

    Parameters:
    ----------
    spot : float
        Initial spot price of the underlying asset.
    strike : float
        Strike price of the option.
    sigma : float
        Volatility of the underlying asset (annualized).
    T : float
        Time to maturity (in years).
    r : float
        Risk-free interest rate (annualized).
    M : int, optional (default=5000)
        Number of Monte Carlo simulation paths.
    N : int, optional (default=5000)
        Number of time steps per path.
    option_type : str, optional (default="call")
        Type of option: "call" or "put".

    Returns:
    -------
    float
        Estimated price of the arithmetic Asian option.

    Notes:
    -----
    The payoff is based on the arithmetic average of the asset prices over time.
    The asset follows a Geometric Brownian Motion (GBM) process.
    """

    # Basic data input check

    if option_type not in ("call", "put"):
        raise ValueError("Please select an option from `call` or `put.")

    # Calculating the time discritisation
    dt = T / N

    # Initialising the spot grid
    spot_grid = np.zeros((M, N + 1))  # N + 1 as N = 0 is the spot price itself

    # Calculating the drift term
    drift = (r - 0.5 * sigma**2) * dt

    # Calculating the diffusion term
    diffusion = sigma * np.random.standard_normal(size=(M, N)) * np.sqrt(dt)

    # Initializing the first time step as spot price
    spot_grid[:, 0] = spot

    # Populating the grid using efficient numpy vectorisation
    spot_grid[:, 1:] = spot * np.cumprod(np.exp(drift + diffusion), axis=1)

    # Calculating the average stock price over the grid
    average_price = np.mean(spot_grid, axis=1)

    # Calculating the option payoff
    payoff = (
        np.maximum(average_price - strike, 0)
        if option_type == "call"
        else np.maximum(strike - average_price, 0)
    )

    # Calculating the mean discounted payoff
    required_payoff = np.exp(-r * T) * np.mean(payoff)

    return required_payoff


if __name__ == "__main__":
    SPOT = 100
    STRIKE = 110
    T = 1
    RISK_FREE = 0.045
    SIGMA = 0.25

    # Calculating the option price
    option_price = arithmetic_asian_option(
        spot=SPOT, strike=STRIKE, sigma=SIGMA, T=T, r=RISK_FREE
    )

    print(option_price)

File: exotic-options/asian_options/monte-carlo-method/arithmetic_asian.cpp
/**
 * @file arithmetic_asian.cpp
 * @brief Monte Carlo simulation for pricing arithmetic Asian options under
 * the Geometric Brownian Motion (GBM) model with variance reduction.
 *
 * This script estimates the fair value of arithmetic Asian options using
 * a Monte Carlo approach. It supports both call and put options, and
 * includes a variance reduction technique to improve accuracy.
 *
 * @author Harsh Parikh
 * @date 31st May 2025
 */

#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

/**
 * @brief Prices an arithmetic Asian option via Monte Carlo simulation
 * under the Geometric Brownian Motion (GBM) framework.
 *
 * @param spot         Spot price of the underlying asset
 * @param strike       Strike price of the option
 * @param sigma        Annualized volatility of the asset
 * @param T            Time to maturity (in years)
 * @param r            Risk-free interest rate
 * @param M            Number of Monte Carlo simulation paths
 * @param N            Number of time steps per path
 * @param option_type  true for call option, false for put option
 * @return             Estimated option price
 */

long double calculatePrice(double spot,
                           double strike,
                           double sigma,
                           double T,
                           double r,
                           int M = 5000,
                           int N = 5000,
                           bool option_type = true)
{
    // Calculating the time discretisation
    double dt = T / N;

    // Initialising the spot grid
}